avr-hal-generic/src/i2c.rs

```diff
@@ -254,26 +236,32 @@ macro_rules! impl_twi_i2c {
                 sda: $sdamod::$SDA<$crate::port::mode::Input<$crate::port::mode::Floating>>,
                 scl: $sclmod::$SCL<$crate::port::mode::Input<$crate::port::mode::Floating>>,
                 speed: u32,
-            ) -> $I2c<CLOCK, $crate::i2c::I2cFloating> {
-                // Calculate TWBR
-                let twbr = ((CLOCK::FREQ / speed) - 16) / 2;
-                p.$twbr.write(|w| unsafe { w.bits(twbr as u8) });
-                // Disable prescaler
-                p.$twsr.write(|w| w.$twps().prescaler_1());
-
-                $I2c {
+            ) -> [<$I2c Master>]<CLOCK, $crate::i2c::I2cFloating> {
+                let mut i2c = [<$I2c Master>] {
                     p,
                     sda,
                     scl,
                     _clock: ::core::marker::PhantomData,
-                }
+                };
+
+                i2c.set_speed(speed);
+                i2c
             }
         }
 
-        impl<CLOCK, M> $I2c<CLOCK, M>
+        impl<CLOCK, M> [<$I2c Master>]<CLOCK, M>
         where
             CLOCK: $crate::clock::Clock,
         {
+            fn set_speed(&mut self, speed: u32){
+                // Calculate TWBR register value
+                let twbr = ((CLOCK::FREQ / speed) - 16) / 2;
```

Comment:
Is there any idea to extend and use https://github.com/Rahix/avr-hal/blob/master/avr-hal-generic/src/usart.rs#L7

Reply:
Hmm, serial baudrate and I2C clock speed are not really related concepts so re-using your code for USART does not quite make sense here.